src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -8,16 +8,128 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Diagnostics
 {
+    internal sealed class AsyncQueue<TElement>
+    {
+        private readonly Channel<TElement> _channel = Channel.CreateUnbounded<TElement>();
+        private readonly TaskCompletionSource<bool> _whenCompleted = new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        /// <summary>
+        /// Gets a task that transitions to a completed state when <see cref="Complete"/> or
+        /// <see cref="TryComplete"/> is called.  This transition will not happen synchronously.
+        /// 
+        /// This Task will not complete until it has completed all existing values returned
+        /// from <see cref="DequeueAsync"/>.
+        /// </summary>
```

Comment:
i copied hte comments untouched.

Reply:
note that this comment seems like a complete lie.  specifically it says "This Task will not complete until it has completed all existing values returned from DequeueAsync".  However, the original semantics don't seem to be that at all.  instead, when CompleteCore is called, it just transitions thsi TCS to the completed state.  so there's no waiting on all existing values being returned.